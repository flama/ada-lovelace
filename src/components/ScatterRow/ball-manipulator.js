const radius = 5

let clearNodes = node => {
  while(node.lastChild) {
    node.removeChild(node.childNodes[0])
  }
}

const Ball = {
  grow(target) {
    if(target.classList.contains('faded')) return
    target.setAttribute('r', radius*2.26)
    target.classList.add('growing')

    let balls = document.getElementsByClassName('d3-point')

    for(let i=0; i<balls.length; ++i)
    {
      if(balls[i] === target) continue

      balls[i].classList.add('shrinking')
      balls[i].setAttribute('r', radius*2/3)
    }

    window.wikiminaTime = setTimeout(() => {
      target.classList.remove('growing')
      target.setAttribute('r', radius*2)
    }, 160)
  },

  shrink(target) {
    if(window.wikiminaTime) {
      clearTimeout(window.wikiminaTime)
      window.wikiminaTime = undefined
    }

    let balls = document.getElementsByClassName('d3-point')
    for(let i=0; i<balls.length; ++i)
    {
      balls[i].classList.remove('growing')
      balls[i].classList.remove('shrinking')
      if(!balls[i].classList.contains('faded'))
        balls[i].setAttribute('r', radius)
    }
  },

  open(target, data) {
    let bubble = document.getElementById('details-bubble')
    let balls = document.getElementsByClassName('d3-point')

    for(let i=0; i<balls.length; ++i)
    {
      balls[i].setAttribute('r', radius*2/3)
      if(balls[i] === bubble) continue

      balls[i].classList.add('shrinking')
      balls[i].classList.add('faded')
    }
    bubble.classList.remove('show')

    setTimeout(() => {
      let rect = target.getBoundingClientRect()
      let plot = document.getElementsByClassName('scatter-plot').item(0).getBoundingClientRect()

      let name = document.createTextNode(data.Name)
      let nameContainer = bubble.getElementsByClassName('name').item(0)
      clearNodes(nameContainer)
      nameContainer.appendChild(name)

      let capitalizedOccupation = data.Occupation[0].toUpperCase() + data.Occupation.slice(1)
      let description = document.createTextNode(`${capitalizedOccupation} from ${data.Country}`)
      let descriptionContainer = bubble.getElementsByClassName('description').item(0)
      clearNodes(descriptionContainer)
      descriptionContainer.appendChild(description)

      let link = bubble.getElementsByClassName('external').item(0)
      link.setAttribute('href', data.Informations)

      bubble.style.top = `${rect.top - plot.top + 5.5}px`
      bubble.style.left = `${rect.left - plot.left + 5.5}px`
      bubble.classList.add('show')
    }, 200)
  }
}

export default Ball
