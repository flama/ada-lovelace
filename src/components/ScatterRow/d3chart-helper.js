import * as d3 from 'd3'

const radius = 10

export default
class d3ChartHelper {

  constructor(el, props, state) {
    let svg = d3.select(el).append('svg')
      .attr('class', 'd3 scatter-row-svg')
      .attr('width', props.width)
      .attr('height', props.height)

    this.g = svg.append('g')
      .attr('class', 'd3-points')

    this.update(el, state)
  }

  update(el, state) {
    let scales = this._scales(el, state.domain)
    this._drawPoints(el, this._format(state.data, scales))
  }

  destroy(el){}

  _format = (data, scales) => data.map(value => {
    return {
      x: scales.x(value.Born),
      y: scales.y(Math.random()),
      extended: value
    }
  })

  _drawPoints = (el, data) => {
    let simulation = d3.forceSimulation(data)
      .force("collide", d3.forceCollide(radius + 1))
      .stop()

    for(let i=0; i<120; ++i) simulation.tick()

    let cell = this.g.append("g")
      .selectAll("g")
      .data(data)
      .enter()
      .append("circle")
      .attr("class", "d3-point")
      .attr("r", radius)
      .attr("cx", d => d.x)
      .attr("cy", d => d.y)
      .attr("onmouseenter", `window.wikiminaGrow(evt.target)`)
      .attr("onmouseleave", `window.wikiminaShrink(evt.target)`)
      .attr("onclick", d => `window.wikiminaOpenBubble(evt.target, ${JSON.stringify(d.extended)})`)

    window.wikiminaGrow = target => {
      target.setAttribute('r', radius*2.26)
      target.classList.add('growing')

      let balls = document.getElementsByClassName('d3-point')

      for(let i=0; i<balls.length; ++i)
      {
        if(balls[i] === target) continue

        balls[i].classList.add('shrinking')
        balls[i].setAttribute('r', radius*2/3)
      }

      window.wikiminaTime = setTimeout(() => {
        target.classList.remove('growing')
        target.setAttribute('r', radius*2)
      }, 160)
    }

    window.wikiminaShrink = target => {
      if(window.wikiminaTime) {
        clearTimeout(window.wikiminaTime)
        window.wikiminaTime = undefined
      }

      let balls = document.getElementsByClassName('d3-point')
      for(let i=0; i<balls.length; ++i)
      {
        balls[i].classList.remove('growing')
        balls[i].classList.remove('shrinking')
        if(!balls[i].classList.contains('faded'))
          balls[i].setAttribute('r', radius)
      }
    }

    window.wikiminaOpenBubble = (target, data) => {
      let bubble = document.getElementById('details-bubble')
      let balls = document.getElementsByClassName('d3-point')
      for(let i=0; i<balls.length; ++i)
      {
        if(balls === bubble) continue

        balls[i].classList.add('shrinking')
        balls[i].classList.add('faded')
      }
      bubble.classList.remove('show')

      setTimeout(() => {
        let rect = target.getBoundingClientRect()
        let plot = document.getElementsByClassName('scatter-plot').item(0).getBoundingClientRect()

        let name = document.createTextNode(data.Name)
        let nameContainer = bubble.getElementsByClassName('name').item(0)
        clearNodes(nameContainer)
        nameContainer.appendChild(name)

        let capitalizedOccupation = data.Occupation[0].toUpperCase() + data.Occupation.slice(1)
        let description = document.createTextNode(`${capitalizedOccupation} from ${data.Country}`)
        let descriptionContainer = bubble.getElementsByClassName('description').item(0)
        clearNodes(descriptionContainer)
        descriptionContainer.appendChild(description)

        let link = bubble.getElementsByClassName('external').item(0)
        link.setAttribute('href', data.Informations)

        bubble.style.top = `${rect.top - plot.top + 20}px`
        bubble.style.left = `${rect.left - plot.left + 20}px`
        bubble.classList.add('show')
      }, 200)
    }

    window.wikiminaCloseBubble = () => {
      let bubble = document.getElementById('details-bubble')
      let balls = document.getElementsByClassName('d3-point')

      for(let i=0; i<balls.length; ++i)
      {
        balls[i].classList.remove('shrinking')
        balls[i].classList.remove('faded')
        balls[i].setAttribute('r', radius)
      }

      bubble.classList.remove('show')
    }

    function clearNodes(node) {
      while(node.lastChild) {
        node.removeChild(node.childNodes[0])
      }
    }

    simulation.on("tick", () => {
      cell
        .attr("cx", d => d.x)
        .attr("cy", d => d.y)
    })
  }

  _scales(el, domain) {
    if(!domain) return null

    let width = el.offsetWidth
    let height = el.offsetHeight

    let x = d3.scaleLinear()
      .range([0, width])
      .domain(domain.x)

    let y = d3.scaleLinear()
      .range([height, 0])
      .domain([-.75, 1.75])

    return { x, y }
  }
}
