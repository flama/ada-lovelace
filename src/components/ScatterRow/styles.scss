@import "../base";

.scatter-row {
  display: flex;
  align-items: center;
  overflow: hidden;

  flex-grow: 1;

  transition: all 0.25s ease-in-out;

  .title {
    display: inline-block;
    height: 100%;
    width: 12vw;

    padding-right: 1em;
    padding-left: 0.5em;

    margin-bottom: -10vh;

    color: $medium;

    h2 {
      font-weight: normal;
    }
  }

  .graph {
    display: inline-block;
    width: 100%;

    .d3-point {
      // fill: $selected;
      fill: white;
      cursor: pointer;
      transition: all 0.2s linear;
      &:hover {
        transition-timing-function: cubic-bezier(.25,5,1,5);
      }
    }
  }

  &:not(.active) {
    flex-grow: 0;
    display: block;
    height: 0;
    overflow: hidden;
  }
}
