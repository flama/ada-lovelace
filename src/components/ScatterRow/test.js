import React from 'react'
import { shallow } from 'enzyme'
import toJSON from 'enzyme-to-json'

import ScatterRow from '../ScatterRow'

const state = {
  domain: {
    x: [1800, 2020]
  },
  data: [{ Born: 1850 }, { Born: 2017 }, { Born: 2000 }],
  test: true
}

describe('ScatterRow', () => {
  let scatterRow

  beforeEach(() => {
    scatterRow = shallow(
      <ScatterRow title="potatoes"
        data={ state.data }
        domain={ state.domain }
        id="obert"
      />
    )
  })

  it('renders correctly', () => {
    expect(toJSON(scatterRow)).toMatchSnapshot()
  })

  it('updates nicely', () => {
    scatterRow.setProps({ title: "baked", ...state })
    expect(toJSON(scatterRow)).toMatchSnapshot()
  })
})
