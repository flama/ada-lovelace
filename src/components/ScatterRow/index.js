import React, { Component } from 'react'
import ReactDOM from 'react-dom'
import d3chartHelper from './d3chart-helper'

import './styles.scss'

class ScatterRow extends Component {

  componentDidMount = () => {
    this.d3Chart = new d3chartHelper(this.graph,{
      width: "100%",
      height: "20vh"
    }, this.getChartState())
  }

  componentDidUpdate = () => {
    let el = ReactDOM.findDOMNode(this)
    this.d3Chart.update(el,this.getChartState())
  }

  getChartState = () => {
    return {
      data: this.reduceDataToArray(this.props.data),
      domain: this.props.domain
    }
  }

  reduceDataToArray = data => {
    return Object.keys(data)
      .reduce((accumulator, key) => {
        return accumulator.concat(data[key])
      }, [])
      .filter(value => value)
  }

  componentWillUnmount = () => {
    let el = ReactDOM.findDOMNode(this)
    this.d3Chart.destroy(el)
  }

  render() {
    return (
      <div className="scatter-row">
        <div className="title" onClick={ () => this.props.open() }><h2>{ this.props.title }</h2></div>
        <div className="graph" ref={ graph => { this.graph = graph } }></div>
      </div>
    )
  }
}

ScatterRow.defaultProps = {
  open: _=>_
}

ScatterRow.propTypes = {
  data: React.PropTypes.object,
  domain: React.PropTypes.object.isRequired,
  open: React.PropTypes.func
}

export default ScatterRow
