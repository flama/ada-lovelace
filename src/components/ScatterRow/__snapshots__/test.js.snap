// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ScatterRow renders correctly 1`] = `
<div
  className="scatter-row "
>
  <div
    className="title"
  >
    <h2>
      potatoes
    </h2>
  </div>
  <div
    className="graph"
  />
</div>
`;

exports[`ScatterRow updates nicely 1`] = `
<div
  className="scatter-row "
>
  <div
    className="title"
  >
    <h2>
      baked
    </h2>
  </div>
  <div
    className="graph"
  />
</div>
`;
